src/prompt/functions/ai-chain.ts

```diff
@@ -0,0 +1,178 @@
+import pMap from 'p-map';
+import { z, type ZodType } from 'zod';
+import type { Model } from '../../model/types.js';
+import {
+  Msg,
+  getErrorMsg,
+  type Prompt,
+  extractZodObject,
+  AbortError,
+} from '../index.js';
+import { type Prettify } from '../../utils/helpers.js';
+
+/**
+ * Creates a chain of chat completion calls that can be invoked as a single
+ * function. It is meant to simplify the process of resolving tool calls
+ * and optionally adding validation to the final result.
+ *
+ * The returned function will invoke the `chatModel` up to `maxCalls` times,
+ * resolving any tool calls to the included `functions` and retrying if
+ * necessary up to `maxRetries`.
+ *
+ * You must pass either `params.messages` or `prompt` to provide the initial
+ * messages to the chat completion model (passing both will throw an error).
+ *
+ * The chain ends when a non-tool call is returned, and the final result can
+ * optionally be validated against a Zod schema, which defaults to a `string`.
+ *
+ * To prevent possible infinite loops, the chain will throw an error if it
+ * exceeds `maxCalls` (`maxCalls` is expected to be >= `maxRetries`).
+ *
+ * Note that `createAIChain` takes in a `functions` array of `AIFunction`
+ * objects created by `createAIFunction`, as the two utility functions are
+ * meant to be used together.
+ */
+export function createAIChain<
+  Params extends Prompt.ChainParams = void,
+  Result extends Prompt.ChainResult = string
+>(
+  {
+    chatModel,
+    schema,
+    params,
+    prompt,
+    functions,
+    maxCalls = 5,
+    maxRetries = 3,
+    toolCallConcurrency = 8,
+  }: {
+    chatModel: Model.Chat.Model;
+    schema?: ZodType<Result>;
+    params?: Omit<
+      Prettify<Model.Chat.Run & Model.Chat.Config>,
+      'functions' | 'tools' | 'function_call' | 'tool_choice'
+    >;
+    prompt?: Prompt.Template<Params>;
```

Comment:
what do you think of something like `input` or `content`. from my use of this in other code, its usually a message directly from the user, or raw data that you are extracting things from.